das/daser.go

```diff
@@ -90,5 +148,97 @@ func (d *DASer) sampling(ctx context.Context, sub header.Subscription) {
+type catchUpJob struct {
+	from, to int64
+}
+
+// catchUpManager manages catch-up jobs, performing them one at a time, exiting
+// only once context is canceled and storing latest DASed checkpoint to disk.
+func (d *DASer) catchUpManager(ctx context.Context, checkpoint int64) {
+	defer func() {
+		// store latest DASed checkpoint to disk here to ensure that if DASer is not yet
+		// fully caught up to network head, it will resume DASing from this checkpoint
+		// up to current network head
+		// TODO @renaynay: Implement Share Cache #180 to ensure no duplicate DASing over same
+		//  header
+		if err := storeCheckpoint(d.cstore, checkpoint); err != nil {
+			log.Errorw("storing checkpoint to disk", "height", checkpoint, "err", err)
+		}
+		// signal that all catchUp routines have finished
+		close(d.catchUpDn)
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case job := <-d.jobsCh:
+			// perform catchUp routine
+			height, err := d.catchUp(ctx, job)
+			// update checkpoint before checking error as the height
+			// value from catchUp represents the last successfully DASed height,
+			// regardless of the routine's success
+			checkpoint = height
+			// exit routine if a catch-up job was unsuccessful
+			if err != nil {
+				log.Errorw("catch-up routine failed", "attempted range (from, to)", job.from,
+					job.to, "last successfully sampled height", height)
+				log.Warn("IN ORDER TO CONTINUE SAMPLING OVER PAST HEADERS, RE-START THE NODE")
+				return
+			}
+		}
+	}
+}
+
+// catchUp starts a sampling routine for headers starting at the next header
+// after the `from` height and exits the loop once `to` is reached. (from:to]
+func (d *DASer) catchUp(ctx context.Context, job *catchUpJob) (int64, error) {
+	routineStartTime := time.Now()
+	log.Infow("sampling past headers", "from", job.from, "to", job.to)
+
+	// start sampling from height at checkpoint+1 since the
+	// checkpoint height is DASed by broader sample routine
+	for height := job.from + 1; height <= job.to; height++ {
+		h, err := d.getter.GetByHeight(ctx, uint64(height))
+		if err != nil {
+			if err == context.Canceled {
+				// report previous height as the last successfully sampled height and
+				// error as nil since the routine was ordered to stop
+				return height - 1, nil
+			}
+
+			log.Errorw("failed to get next header", "height", height, "err", err)
+			// report previous height as the last successfully sampled height
+			return height - 1, err
+		}
+
+		startTime := time.Now()
+
+		err = d.da.SharesAvailable(ctx, h.DAH)
+		if err != nil {
+			if err == context.Canceled {
+				// report previous height as the last successfully sampled height and
+				// error as nil since the routine was ordered to stop
+				return height - 1, nil
+			}
+			log.Errorw("sampling failed", "height", h.Height, "hash", h.Hash(),
+				"square width", len(h.DAH.RowsRoots), "data root", h.DAH.Hash(), "err", err)
+			// continue sampling
```

Comment:
For some reason I thought you PR makes the code to return here, especially with warning that tells that you need to restart the node to continue sampling, but the return happens only in case the header is not available. I think it is ok to continue here and all the retries and etc should happen inside SharesAvailable. 

Also BEFP handling should probably happen inside. So like FullAvailability generates it and LighAvailbaility rcvs it and signals to DASer via special error returned from SharesAvailable or something like this. This can also help with case of restarting of LightNode, so once DASer asks LightAvailablity, it will check if there are any known BEFP(either newly rcvs or stored) and return an error on which DASer will terminate.